File: Convex Hull/Brute-force/CHBF.py
def orientation(p, q, r):
    """Find the orientation of the triplet (p, q, r)."""
    val = (q[1] - p[1]) * (r[0] - q[0]) - (q[0] - p[0]) * (r[1] - q[1])
    if val == 0:
        return 0 # collinear
    elif val > 0:
        return 1 # clockwise
    else:
        return 2 # counterclockwise

def convex_hull(points):
    """Find the convex hull of a set of points."""
    n = len(points)
    if n < 3:
        return points

    hull = []
    for i in range(n):
        for j in range(i+1, n):
            # check if (i, j) makes up an edge of the convex hull
            left = []
            right = []
            for k in range(n):
                if k != i and k != j:
                    if orientation(points[i], points[j], points[k]) == 2:
                        left.append(k)
                    else:
                        right.append(k)
            if not left or not right:
                hull.append((points[i], points[j]))

    # construct the convex hull from the edges
    ch = []
    for (p, q) in hull:
        if p not in ch:
            ch.append(p)
        if q not in ch:
            ch.append(q)
    ch.sort(key=lambda x: (x[0], x[1]))

    return ch

File: Convex Hull/Brute-force/CHBF.java
import java.util.*;

public class ConvexHull {
    public static void main(String[] args) {
        // Example set of points
        int[][] points = {{0, 0}, {1, 1}, {2, 2}, {2, 0}, {3, 3}, {4, 4}, {4, 0}};

        // Find the convex hull
        List<int[]> hull = bruteForceConvexHull(points);

        // Print the convex hull
        System.out.println("Convex Hull:");
        for (int[] point : hull) {
            System.out.println(Arrays.toString(point));
        }
    }

    public static List<int[]> bruteForceConvexHull(int[][] points) {
        // Initialize the convex hull as an empty list
        List<int[]> hull = new ArrayList<>();

        // Check each pair of points to see if it makes up an edge of the convex hull
        for (int i = 0; i < points.length; i++) {
            for (int j = i + 1; j < points.length; j++) {
                int[] p1 = points[i];
                int[] p2 = points[j];

                // Find the equation of the line between p1 and p2: y = mx + b
                double m = (p2[1] - p1[1]) / (double) (p2[0] - p1[0]);
                double b = p1[1] - m * p1[0];

                // Check if all other points are on one side of the line
                boolean onOneSide = true;
                for (int[] p : points) {
                    if (p == p1 || p == p2) {
                        continue;
                    }
                    double y = m * p[0] + b;
                    if (p[1] < y) {
                        onOneSide = false;
                        break;
                    }
                }

                // If all other points are on one side of the line, add it to the convex hull
                if (onOneSide) {
                    if (!hull.contains(p1)) {
                        hull.add(p1);
                    }
                    if (!hull.contains(p2)) {
                        hull.add(p2);
                    }
                }
            }
        }

        return hull;
    }
}